lint/capability_fields_analyser.go

```diff
@@ -20,35 +20,69 @@ package lint
 
 import (
 	"github.com/onflow/cadence/runtime/ast"
+	"github.com/onflow/cadence/runtime/common"
 	"github.com/onflow/cadence/tools/analysis"
 )
 
-func DetectCapabilityType(typeToCheck ast.Type) bool {
+func DetectCapabilityType(typeToCheck ast.Type, structWithPubCapabilities map[string]struct{}) bool {
 	const capabilityTypeName = "Capability"
 	switch downcastedType := typeToCheck.(type) {
 	case *ast.NominalType:
-		return downcastedType.Identifier.Identifier == capabilityTypeName
+		_, found := structWithPubCapabilities[downcastedType.Identifier.Identifier]
+		return downcastedType.Identifier.Identifier == capabilityTypeName || found
 	case *ast.OptionalType:
-		return DetectCapabilityType(downcastedType.Type)
+		return DetectCapabilityType(downcastedType.Type, structWithPubCapabilities)
 	case *ast.VariableSizedType:
-		return DetectCapabilityType(downcastedType.Type)
+		return DetectCapabilityType(downcastedType.Type, structWithPubCapabilities)
 	case *ast.ConstantSizedType:
-		return DetectCapabilityType(downcastedType.Type)
+		return DetectCapabilityType(downcastedType.Type, structWithPubCapabilities)
 	case *ast.DictionaryType:
-		return DetectCapabilityType(downcastedType.KeyType) || DetectCapabilityType(downcastedType.ValueType)
+		return DetectCapabilityType(downcastedType.KeyType, structWithPubCapabilities) || DetectCapabilityType(downcastedType.ValueType, structWithPubCapabilities)
 	case *ast.FunctionType:
 		return false
 	case *ast.ReferenceType:
-		return DetectCapabilityType(downcastedType.Type)
+		return DetectCapabilityType(downcastedType.Type, structWithPubCapabilities)
 	case *ast.RestrictedType:
 		return false
 	case *ast.InstantiationType:
-		return DetectCapabilityType(downcastedType.Type)
+		return DetectCapabilityType(downcastedType.Type, structWithPubCapabilities)
 	default:
 		panic("Unknown type")
 	}
 }
 
+func CollectStructsWithPublicCapabilities(inspector *ast.Inspector) (map[string]struct{}, map[ast.Identifier]struct{}) {
+	structWithPubCapabilities := make(map[string]struct{})
+	fieldsInStruct := make(map[ast.Identifier]struct{})
+	inspector.Preorder(
+		[]ast.Element{(*ast.CompositeDeclaration)(nil)},
+		func(element ast.Element) {
+			switch declaration := element.(type) {
+			case *ast.CompositeDeclaration:
+				{
+					if declaration.CompositeKind != common.CompositeKindStructure {
+						return
+					}
+					for _, d := range declaration.Members.Declarations() {
+						field, ok := d.(*ast.FieldDeclaration)
+						if !ok {
+							return
+						}
+						if field.Access != ast.AccessPublic {
+							return
+						}
+						if field.Access == ast.AccessPublic && DetectCapabilityType(field.TypeAnnotation.Type, structWithPubCapabilities) {
```

Comment:
if field.Access == ast.AccessPublic will allways be false because you check it before, isn't it?

Reply:
thx, you are correct its superfluous, I had it on my local machine's branch too.